### PR TITLE
Add support to set and reset volume options

### LINKED ISSUE #76: Volume Option Set/Reset
Option Set

```
kadalu volume set <name> <optname> <optvalue> ...
```

Option Reset

```
kadalu volume reset <name> <optname1> <optname2>...
```

- Notification should be sent to affected Server nodes
- Notification should be sent to affected Client nodes(TBD or as a separate issue)

### PR DESCRIPTION
**Pull request by vatsa287**
Fixes: #76 
Signed-off-by: Shree Vatsa N <vatsa@kadalu.io>

### CODE REVIEW

File: clients/crystal/src/volumes.cr [hunk 1]
```diff
@@ -99,6 +99,38 @@ module MoanaClient
       start_stop_volume("stop")
     end
 
+    def set(volume_options : String)
```
Comment (aravindavk) [inline clients/crystal/src/volumes.cr]: Accept `Hash(String, String)` here.

File: clients/crystal/src/volumes.cr [hunk 2]
```diff
@@ -99,6 +99,38 @@ module MoanaClient
...
+      url = "#{@client.url}/api/v1/pools/#{@pool_name}/volumes/#{@name}/options/set"
+
+      puts "in set", volume_options
+      response = MoanaClient.http_post(
+        url,
+        ({"volume_options": volume_options}).to_json,
```
Comment (aravindavk) [inline clients/crystal/src/volumes.cr]: directly send input options. `volume_options.to_json`

File: clients/crystal/src/volumes.cr [hunk 3]
```diff
@@ -99,6 +99,38 @@ module MoanaClient
...
+        headers: @client.auth_header
+      )
+      if response.status_code == 201
+        MoanaTypes::Volume.from_json(response.body)
+      else
+        MoanaClient.error_response(response)
+      end
+    end
+
+    def reset(volume_option_keys : Array)
```
Comment (aravindavk) [inline clients/crystal/src/volumes.cr]: Same here. Accept `Array(String)`

File: clients/crystal/src/volumes.cr [hunk 4]
```diff
@@ -99,6 +99,38 @@ module MoanaClient
...
+      url = "#{@client.url}/api/v1/pools/#{@pool_name}/volumes/#{@name}/options/reset"
+
+      puts "in reset", volume_option_keys
+      response = MoanaClient.http_post(
+        url,
+        ({"volume_option_keys": volume_option_keys}).to_json,
```
Comment (aravindavk) [inline clients/crystal/src/volumes.cr]: directly send the input option keys. `volume_option_keys.to_json`

File: mgr/src/cmds/helpers.cr [hunk 5]
```diff
@@ -125,3 +125,21 @@ def handle_json_error(message, args)
...
+  end
+
+  vol_opts = Hash(String, String).new
+  volume_options.each_with_index do |volume_option, index|
+    if index == volume_options.size - 1
+      break
+    end
+    next unless index % 2 == 0
+    vol_opts[volume_option] = volume_options[index + 1]
+  end
```
Comment (aravindavk) [inline mgr/src/cmds/helpers.cr]: Line 136-142

```
volume_options.each_slice(2) do |opt|
  vol_opts[opt[0]] = opt[1]
end
```

File: mgr/src/cmds/volumes.cr [hunk 6]
```diff
@@ -239,3 +239,38 @@ handler "volume.delete" do |args|
...
+handler "volume.set" do |args|
+  args.pool_name, volume_name = pool_and_volume_name(args.pos_args.size > 0 ? args.pos_args[0] : "")
+
+  volume_options = validate_volume_options(args.pos_args[1..])
+  # without .to_json to conserve HASH.
+  puts "here", volume_options.to_json.to_s
+  puts "here", typeof(volume_options.to_s)
+
+  api_call(args, "Failed to Set options to the Volume") do |client|
+    volume = client.pool(args.pool_name).volume(volume_name).set(volume_options.to_json.to_s)
```
Comment (aravindavk) [inline mgr/src/cmds/volumes.cr]:
```
volume = client.pool(args.pool_name).volume(volume_name).set(volume_options)
```

File: mgr/src/server/plugins/volume_option_set_reset.cr [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  volume_name = env.params.url["volume_name"]
+
+  puts pool_name
+
+  next forbidden(env) unless Datastore.maintainer?(env.user_id, pool_name)
+
+  # move this after validations:::
+  # puts (env.params.json.fetch("volume_options", "").as(String))
+  # puts (env.params.body["volume_options"])
+  volume_options = env.params.json["volume_options"].as(String)
```
Comment (aravindavk) [inline mgr/src/server/plugins/volume_option_set_reset.cr]:
```
volume_options = Hash(String, String).from_json(env.request.body.not_nil!)
```

File: mgr/src/server/plugins/volume_option_set_reset.cr [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  volume = Datastore.get_volume(pool_name, volume_name)
+  if volume.nil?
+    halt(env, status_code: 400, response: ({"error": "The Volume(#{volume_name}) doesn't exists"}.to_json))
+  end
+
+  puts "in here"
+  puts typeof(volume.options)
+  puts volume.options
+
+  volume.options = volume.options.merge(Hash(String, String).from_json(volume_options))
```
Comment (aravindavk) [inline mgr/src/server/plugins/volume_option_set_reset.cr]:
```
volume.options = volume.options.merge(volume_options)
```

File: mgr/src/server/plugins/volume_option_set_reset.cr [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  puts volume.options
+
+  # update volume options to DB.
+  Datastore.update_volume_options(volume.id, volume.options.to_json.to_s)
```
Comment (aravindavk) [inline mgr/src/server/plugins/volume_option_set_reset.cr]: `.to_s` is redundant. `to_json` itself convert to string format.

Comment (aravindavk) [reply mgr/src/server/plugins/volume_option_set_reset.cr]: Also move `to_json` inside `update_volume_options` function

File: mgr/src/server/plugins/volume_option_set_reset.cr [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+post "/api/v1/pools/:pool_name/volumes/:volume_name/options/reset" do |env|
+  pool_name = env.params.url["pool_name"]
+  volume_name = env.params.url["volume_name"]
+
+  puts pool_name
+
+  next forbidden(env) unless Datastore.maintainer?(env.user_id, pool_name)
+
+  # move this after validations:::
+  volume_option_keys = env.params.json["volume_option_keys"].as(Array)
```
Comment (aravindavk) [inline mgr/src/server/plugins/volume_option_set_reset.cr]:
```
volume_option_keys = Array(String).from_json(env.request.body.not_nil!)
```

File: mgr/src/server/plugins/volume_option_set_reset.cr [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  end
+
+  volume = Datastore.get_volume(pool_name, volume_name)
+  if volume.nil?
+    halt(env, status_code: 400, response: ({"error": "The Volume(#{volume_name}) doesn't exists"}.to_json))
+  end
+
+  volume.options = volume.options.reject(volume_option_keys)
+  # update volume options to DB.
+  Datastore.update_volume_options(volume.id, volume.options.to_json.to_s)
```
Comment (aravindavk) [inline mgr/src/server/plugins/volume_option_set_reset.cr]: Remove `to_json.to_s`

File: volgen/src/moana_volgen.cr [hunk 12]
```diff
@@ -221,6 +221,7 @@ class Volfile
 
     volumes.each_with_index do |volume, vidx|
       opts = volume.options
+
```
Comment (aravindavk) [inline volgen/src/moana_volgen.cr]: Extra change!

File: mgr/src/server/datastore/volumes.cr [hunk 13]
```diff
@@ -243,6 +245,11 @@ module Datastore
     end
   end
 
+  def update_volume_options(volume_id, options)
+    query = update_query("volumes", ["options"], where: " id = ?")
+    connection.exec(query, options, volume_id)
```
Comment (aravindavk) [inline mgr/src/server/datastore/volumes.cr]:
```
connection.exec(query, options.to_json, volume_id)
```

**Review by aravindavk — CHANGES_REQUESTED** [review]
Remove the Print statements from the code.

File: clients/crystal/src/volumes.cr (hunk 1 above)
Comment (vatsa287) [reply clients/crystal/src/volumes.cr]: done

File: clients/crystal/src/volumes.cr (hunk 2 above)
Comment (vatsa287) [reply clients/crystal/src/volumes.cr]: done

File: clients/crystal/src/volumes.cr (hunk 3 above)
Comment (vatsa287) [reply clients/crystal/src/volumes.cr]: done

**Review by aravindavk — APPROVED** [review]

**Comment by aravindavk** [comment]
Rebase is required.

**Pull request merged**